### DISCUSSION TITLE
NES Salamander (Japan)

### DISCUSSION THREAD

**Discussion by splash5** [sanni/cartreader#657] [Need help]
Is there anyone dump "Salamander (Japan)" successfully?
I always got first bank dumped (repeat for 8 times).
Seems VRC3 can not switch bank well, but it runs no problem on real hardware.
Here is waveform I logged with logic analyzer on real hardware.
![vrc3](https://user-images.githubusercontent.com/2173030/207326145-67ac13f9-2f05-4ced-ad3c-a5ddfc2b9cfd.png)

**Comment by nsx0r** [comment]
can you try replacing the line at https://github.com/sanni/cartreader/blob/master/Cart_Reader/NES.ino#L3015

`write_reg_byte(0xF000, i);`

with this line:

`write_prg_byte(0xF000, i | 8);`

**Reply by splash5** [reply]
I tried `write_prg_byte(0xF000, i | 8);` and it works! Thanks a lot!
Also tested mapper 68 and works well too, save looks not corrupted.

So the problem seems belong to /ROMSEL and M2 switching... (too fast for VRC3 I think)

**Reply by nsx0r** [reply]
possibly a timing issue, or bit 3 must always be set for some reason?
you can test this theory with `write_reg_byte(0xF000, i | 8);` and `write_prg_byte(0xF000, i);`

**Reply by splash5** [reply]
I have tested dumping with `write_prg_byte(0xF000, i);` and no problem at all.
Dumping with `write_reg_byte(0xF000, i | 8);` is not working. So this must be a timing problem.
According to [here](https://github.com/sharkpp/unagi_script_rom/blob/master/vrc3.ud), set bit 3 of $F000 reg will map 0x1c000-0x1cfff to CPU address $c000-$cfff.
But not affected if reading last bank from $8000-$BFFF, so not ORing 0x08 with bank id is ok.
